src/mpi/coll/ireduce/ireduce.c

```diff
@@ -397,6 +397,26 @@ int MPIR_Ireduce(const void *sendbuf, void *recvbuf, int count,
                                       request);
     }
 
+    if (!*request)
+        goto immed_complete;
+
+    MPIR_REQUEST_CS_ENTER(*request);
```

Comment:
We need to ensure that if the request is not complete, the struct members used for copying data back to the user buffer are updated before the request can complete in another thread. Just checking the whether the request is complete would not need the lock, but combined with the member updates, it does. cc @hzhou

Reply:
Would it make any difference if the values of the struct members are updated prior to `MPID_Ireduce` or `MPIR_Iallreduce_impl` call?

Reply:
Yes, but currently the request is not yet created at that stage. Request creation happens within the collective algos.

Reply:
We could move request creation up to this layer, but that would eliminate some (minor) optimizations for immediately complete collective operations.

Reply:
It depends on whether there will be contention. If we are certain that only a single thread will modify and access the data then we don't need a critical section for it. I think the current design there will be two places accessing the buffer data, one here, and the other one potentially in the request completion path, thus a critical section is needed

Reply:
Thank you for explaining.  I forgot about the request being created within the algorithm.

Reply:
I am not sure if this critical section here is enough. What if progress thread kicks in and enter the request completion path before this call enters the critical section?

Reply:
In that case, when here finally enters the critical section, it will find the request is completed, and jump to `immed_complete` -- line 406 below

Reply:
> In that case, when here finally enters the critical section, it will find the request is completed, and jump to `immed_complete` -- line 406 below

Exactly. We still have to check the request status after we've acquired the lock. It also occurs to me since this code is all related to GPU support, we can add a branch to shortcut it in the CPU-only case.